lib/dry/core/memoizable.rb

```diff
@@ -12,7 +12,14 @@ def memoize(*names)
 
         def new(*)
           obj = super
-          obj.instance_eval { @__memoized__ = MEMOIZED_HASH.dup }
+
+          case obj
+          when Object
+            obj.instance_variable_set(:'@__memoized__', MEMOIZED_HASH.dup)
+          else
+            obj.instance_eval { @__memoized__ = MEMOIZED_HASH.dup }
+          end
```

Comment:
@solnic Here's one approach. I was experimenting a bit with refinements prior to this for separating the two `new` implementations but the current implementation of `refine` is too restrictive (by design). 

I'll do some benchmarking later today using this `case/when` approach. Let me know what you think, is there a more elegant solution?

Reply:
I'd say it should all be decided in the `included` hook below since base class cannot be changed

Reply:
What Nikita said *or* maybe this should be set lazily via `def memoized; @__memoized__ ||= MEMOIZED_HASH.dup; end`?

Reply:
Good point with frozen objects. OK so let's handle it in the included hook like Nikita suggested. It'd simply require two separate modules with custom `new` implementations and just extending the class with the right one based on parent class.

Reply:
@solnic Sorry about the delay. Had some personal business I had to take care of over the weekend. I've pushed my new suggestion.